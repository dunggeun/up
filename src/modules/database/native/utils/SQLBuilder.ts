/* eslint-disable no-case-declarations */
import dedent from 'dedent';
import { Logger } from 'src/modules/logger';
import type { DatabaseRecord, Value, WhereConditions } from '../../types';

type QueryType = 'select' | 'count' | 'insert' | 'update' | 'delete';

const TAG = 'SQLBuilder';

export class SQLBuilder {
  private _tableName?: string;
  private _type?: QueryType;
  private _values?: DatabaseRecord;
  private _limit?: number;
  private _orderTarget?: string;
  private _by?: 'desc' | 'asc';
  private _condition?: WhereConditions;

  private stringify(value: Value): string {
    if (typeof value === 'string') {
      return `"${value}"`;
    } else if (value === null) {
      return 'null';
    }
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    return `${value}`;
  }

  table(tableName: string): this {
    this._tableName = tableName;
    return this;
  }

  type(type: QueryType): this {
    this._type = type;
    return this;
  }

  values(values: DatabaseRecord): this {
    this._values = values;
    return this;
  }

  where(condition: WhereConditions): this {
    this._condition = condition;
    return this;
  }

  limit(limit: number): this {
    this._limit = limit;
    return this;
  }

  orderBy(target: string, by: 'desc' | 'asc'): this {
    this._orderTarget = target;
    this._by = by;
    return this;
  }

  build(): string {
    if (!(this._type && this._tableName)) {
      throw Error('invalid query type or table name');
    }

    let query: string;
    const WHERE_CONDITION =
      this._condition &&
      Object.entries(this._condition)
        .map(([column, { symbol, value }]) => {
          return `${column} ${symbol} ${this.stringify(value)}`;
        })
        .join(' AND ');
    const WHERE_STATEMENTS = WHERE_CONDITION ? `WHERE ${WHERE_CONDITION}` : '';

    switch (this._type) {
      case 'select':
        const shouldOrdering = this._orderTarget && this._by;
        const ORDER_BY_STATEMENTS = shouldOrdering
          ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            `ORDER BY ${this._orderTarget!} ${this._by!}`
          : '';

        const LIMIT_CONSTRAINT =
          this._limit !== undefined ? `LIMIT ${this._limit}` : '';

        query = dedent`
          SELECT * FROM ${this._tableName}
          ${WHERE_STATEMENTS}
          ${ORDER_BY_STATEMENTS}
          ${LIMIT_CONSTRAINT};
        `;
        break;

      case 'count':
        query = dedent`SELECT count(*) as count FROM ${this._tableName};`;
        break;

      case 'insert':
        const VALUE_COLUMNS = Object.keys(this._values ?? {}).join(',');
        const VALUES = Object.values(this._values ?? {})
          .map((value) => this.stringify(value))
          .join(',');

        query = dedent`
          INSERT INTO ${this._tableName} (${VALUE_COLUMNS})
          VALUES (${VALUES});
        `;
        break;

      case 'update':
        const SET_STATEMENTS = Object.entries(this._values ?? {})
          .map(([column, value]) => `${column} = ${this.stringify(value)}`)
          .join(',');

        query = dedent`
          UPDATE ${this._tableName}
          SET ${SET_STATEMENTS}
          ${WHERE_STATEMENTS};
        `;
        break;

      case 'delete':
        query = dedent`
          DELETE FROM ${this._tableName}
          ${WHERE_STATEMENTS};
        `;
        break;
    }

    query = query.replace(/[\r\n]/g, ' ');
    Logger.debug(TAG, query);

    return query;
  }
}
