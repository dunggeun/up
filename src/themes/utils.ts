import { Platform } from 'react-native';

export const webFont = (font: string): string => {
  return Platform.select({
    web: `${font}, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, Inter-serif`,
    default: font,
  });
};
