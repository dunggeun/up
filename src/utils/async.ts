export const delay = (duration: number): Promise<void> => {
  return new Promise((resolve) => {
    setTimeout(resolve, duration);
  });
};

/* istanbul ignore next */

export const debounce = <T>(
  callback: (...args: T[]) => void,
  // eslint-disable-next-line @typescript-eslint/no-shadow
  delay: number,
): ((...args: T[]) => void) => {
  let timer: NodeJS.Timeout;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => callback(...args), delay);
  };
};
