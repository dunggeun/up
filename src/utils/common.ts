import { Platform } from 'react-native';
import {
  trigger,
  HapticFeedbackTypes as RNHapticFeedbackTypes,
} from 'react-native-haptic-feedback';

const HAPTIC_FEEDBACK_TYPES = {
  press: Platform.select<RNHapticFeedbackTypes>({
    ios: RNHapticFeedbackTypes.impactLight,
    android: RNHapticFeedbackTypes.virtualKey,
  }),
  buttonPress: Platform.select<RNHapticFeedbackTypes>({
    ios: RNHapticFeedbackTypes.impactMedium,
    android: RNHapticFeedbackTypes.effectClick,
  }),
  buttonLongPress: Platform.select<RNHapticFeedbackTypes>({
    ios: RNHapticFeedbackTypes.impactHeavy,
    android: RNHapticFeedbackTypes.effectHeavyClick,
  }),
} as const;

export type HapticFeedbackTypes = keyof typeof HAPTIC_FEEDBACK_TYPES;

/* istanbul ignore next */
// eslint-disable-next-line @typescript-eslint/no-empty-function
export function noop(): void {}

export function triggerHaptic(type: HapticFeedbackTypes = 'press'): void {
  const feedbackType = HAPTIC_FEEDBACK_TYPES[type];
  feedbackType && trigger(feedbackType);
}
