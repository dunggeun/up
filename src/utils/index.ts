export * from './async';
export * from './common';
export * from './locales';
