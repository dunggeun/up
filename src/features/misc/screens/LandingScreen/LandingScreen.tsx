import React from 'react';
import { styled, View, Image } from 'dripsy';
import { MotiView } from 'moti';
import Logo from 'src/assets/images/logo.png';
import { LANDING_LOGO_SIZE, LANDING_LOGO_MARGIN } from 'src/constants';
import { CommonLayout, Button, H1 } from 'src/designs';
import { t } from 'src/translations';
import type { RootStackProps } from 'src/navigators/RootStack/types';

type LandingScreenProps = RootStackProps<'Landing'>;

const PageTitleArea = styled(View)({
  paddingY: '$04',
});

const LogoArea = styled(View)({
  flex: 1,
  justifyContent: 'center',
  alignItems: 'center',
  paddingBottom: LANDING_LOGO_MARGIN,
});

const LogoImage = styled(Image)({
  width: '100%',
  height: '100%',
});

const scaleLogoViewStyle = {
  width: LANDING_LOGO_SIZE,
  height: LANDING_LOGO_SIZE,
} as const;

export function LandingScreen({ navigation }: LandingScreenProps): JSX.Element {
  const handlePressNextButton = (): void => {
    navigation.navigate('UserRegister');
  };

  return (
    <CommonLayout>
      <CommonLayout.Header />
      <CommonLayout.Body scrollEnabled={false}>
        <PageTitleArea>
          <H1 variant="primary">{t('message.greeting')}</H1>
        </PageTitleArea>
        <LogoArea>
          <MotiView
            animate={{ scale: 1.5 }}
            from={{ scale: 1 }}
            style={scaleLogoViewStyle}
            transition={{
              type: 'spring',
              damping: 2.5,
              stiffness: 120,
            }}
          >
            <LogoImage source={Logo} />
          </MotiView>
        </LogoArea>
      </CommonLayout.Body>
      <CommonLayout.Footer>
        <Button color="$brand" disableLongPress onPress={handlePressNextButton}>
          {t('label.start')}
        </Button>
      </CommonLayout.Footer>
    </CommonLayout>
  );
}
