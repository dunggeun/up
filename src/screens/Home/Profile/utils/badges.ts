import { t } from 'src/translations';
import dummy from 'src/assets/badges/dummy.png';

import type { ImageSourcePropType } from 'react-native';

const BADGE_SET = [[t('badge.0'), dummy]] as const;
const FALLBACK_BADGE = ['', dummy] as const;

export const getBadge = (id: number): { title: string, image: ImageSourcePropType } => {
  const badge = BADGE_SET[id];
  const [title, image] = badge ?? FALLBACK_BADGE;
  return { title, image };
};
