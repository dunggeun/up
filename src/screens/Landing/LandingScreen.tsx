import React, { useEffect } from 'react';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withSpring
} from 'react-native-reanimated';
import { styled, View, Image } from 'dripsy';
import { CommonLayout, Button, H1 } from 'src/designs';
import { LANDING_LOGO_SIZE, LANDING_LOGO_MARGIN } from 'src/constants';
import { t } from 'src/translations';
import Logo from 'src/assets/images/logo.png';

import type { RootStackProps } from 'src/navigators/RootStack/types';

type LandingScreenProps = RootStackProps<'Landing'>;

const PageTitleArea = styled(View)({
  paddingY: '$04',
});

const LogoArea = styled(View)({
  flex: 1,
  justifyContent: 'center',
  alignItems: 'center',
  paddingBottom: LANDING_LOGO_MARGIN,
});

const LogoImage = styled(Image)({
  width: '100%',
  height: '100%',
});

const scaleLogoViewStyle = {
  width: LANDING_LOGO_SIZE,
  height: LANDING_LOGO_SIZE,
} as const;

export function LandingScreen({ navigation }: LandingScreenProps): JSX.Element {
  const scale = useSharedValue(1);
  const animatedLogoViewStyle = useAnimatedStyle(() => ({
    transform: [{ scale: scale.value }],
  }));

  useEffect(() => {
    scale.value = withSpring(1.5, { damping: 2.5, stiffness: 120 });
  }, [scale]);

  const handlePressNextButton = (): void => {
    navigation.navigate('RegisterUser');
  };

  return (
    <CommonLayout>
      <CommonLayout.Header />
      <CommonLayout.Body scrollEnabled={false}>
        <PageTitleArea>
          <H1 variant="primary">{t('message.greeting')}</H1>
        </PageTitleArea>
        <LogoArea>
          <Animated.View style={[scaleLogoViewStyle, animatedLogoViewStyle]}>
            <LogoImage source={Logo} />
          </Animated.View>
        </LogoArea>
      </CommonLayout.Body>
      <CommonLayout.Footer>
        <Button
          color="$brand"
          disableLongPress
          onPress={handlePressNextButton}
        >
          {t('label.start')}
        </Button>
      </CommonLayout.Footer>
    </CommonLayout>
  );
}
