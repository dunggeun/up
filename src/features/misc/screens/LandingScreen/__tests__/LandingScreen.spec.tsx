import React from 'react';
import { render, cleanup } from '@testing-library/react-native';
import { withDripsy } from 'tests/utils';

import { LandingScreen } from '../LandingScreen';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getMockedProps = (): any => ({
  navigation: {
    goBack: jest.fn(),
  },
});

describe('screens/LandingScreen', () => {
  afterEach(() => {
    cleanup();
  });

  describe('렌더링 되었을 때', () => {  
    it('스냅샷이 일치해야 한다', () => {
      const tree = render(
        withDripsy(<LandingScreen {...getMockedProps()} />)
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
