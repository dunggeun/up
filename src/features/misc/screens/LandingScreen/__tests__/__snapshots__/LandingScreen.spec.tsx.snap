// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screens/LandingScreen 렌더링 되었을 때 스냅샷이 일치해야 한다 1`] = `
<View
  style={
    [
      {},
      {
        "flex": 1,
      },
      undefined,
      {
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
      },
      {},
      {},
    ]
  }
>
  <View
    style={
      [
        {
          "alignSelf": "center",
          "backgroundColor": "#ffffff",
          "flex": 1,
          "marginHorizontal": "auto",
          "maxWidth": 600,
          "paddingHorizontal": 16,
          "width": "100%",
        },
        undefined,
        {},
      ]
    }
  >
    <View
      style={
        [
          {},
          {
            "marginHorizontal": -16,
            "zIndex": 3,
          },
          undefined,
          {},
          {},
        ]
      }
    >
      <View
        style={
          [
            {},
            {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "flexDirection": "row",
              "height": 56,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "position": "relative",
            },
            undefined,
            {},
            {},
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              {
                "bottom": 0,
                "elevation": 3,
                "height": 16,
                "left": 0,
                "marginBottom": -16,
                "position": "absolute",
                "right": 0,
                "zIndex": 3,
              },
              undefined,
              {},
              {},
            ]
          }
        >
          <RNSVGSvgView
            align="none"
            bbHeight="100%"
            bbWidth="100%"
            focusable={false}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                {
                  "flex": 0,
                  "height": "100%",
                  "width": "100%",
                },
              ]
            }
            vbHeight={100}
            vbWidth={100}
          >
            <RNSVGGroup
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
            >
              <RNSVGDefs>
                <RNSVGLinearGradient
                  gradient={
                    [
                      0,
                      -1,
                      1,
                      16777215,
                    ]
                  }
                  gradientTransform={
                    [
                      6.123233995736766e-17,
                      1,
                      -1,
                      6.123233995736766e-17,
                      0,
                      0,
                    ]
                  }
                  gradientUnits={1}
                  name="g"
                  x1="0"
                  x2="100"
                  y1="0"
                  y2="0"
                />
              </RNSVGDefs>
              <RNSVGRect
                fill={
                  {
                    "brushRef": "g",
                    "type": 1,
                  }
                }
                height="100%"
                propList={
                  [
                    "fill",
                  ]
                }
                width="100%"
                x="0"
                y="0"
              />
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
        <View
          style={
            [
              {},
              {
                "height": 30,
                "width": 30,
              },
              undefined,
              {},
              {},
            ]
          }
        />
        <Text
          accessibilityHint="Title"
          accessibilityLabel="Empty"
          accessibilityRole="header"
          style={
            [
              {
                "fontSize": 48,
                "fontWeight": "bold",
                "marginVertical": 26.56,
              },
              [
                {
                  "fontFamily": "BMJUA",
                  "fontSize": 24,
                  "marginVertical": 0,
                  "paddingBottom": 4,
                },
                {
                  "fontWeight": "normal",
                },
                undefined,
                {
                  "color": "#2e2e2e",
                },
              ],
            ]
          }
        />
        <View
          style={
            [
              {},
              {
                "height": 30,
                "width": 30,
              },
              undefined,
              {},
              {},
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        [
          {},
          {
            "flex": 1,
          },
          {},
        ]
      }
    >
      <View
        style={
          [
            {},
            {
              "flex": 1,
              "gap": 16,
            },
            undefined,
            {},
            {},
          ]
        }
      >
        <View
          style={
            [
              {},
              {
                "paddingVertical": 16,
              },
              undefined,
              {},
              {},
            ]
          }
        >
          <Text
            accessibilityRole="header"
            style={
              [
                {
                  "fontSize": 64,
                  "fontWeight": "bold",
                  "marginVertical": 21.44,
                },
                [
                  {
                    "color": "#2e2e2e",
                    "fontFamily": "BMJUA",
                    "fontSize": 32,
                    "marginVertical": 0,
                    "paddingBottom": 4,
                  },
                  {
                    "fontWeight": "normal",
                  },
                  undefined,
                  {},
                ],
              ]
            }
          >
            Welcome!
Take the Up Challenge
          </Text>
        </View>
        <View
          style={
            [
              {},
              {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
                "paddingBottom": 100,
              },
              undefined,
              {},
              {},
            ]
          }
        >
          <View
            animate={
              {
                "scale": 1.5,
              }
            }
            from={
              {
                "scale": 1,
              }
            }
            style={
              [
                {
                  "height": 180,
                  "width": 180,
                },
                {
                  "transform": [
                    {
                      "scale": 1,
                    },
                  ],
                },
              ]
            }
            transition={
              {
                "damping": 2.5,
                "stiffness": 120,
                "type": "spring",
              }
            }
          >
            <Image
              source={
                {
                  "testUri": "../../../src/assets/images/logo.png",
                }
              }
              style={
                [
                  {},
                  {
                    "height": "100%",
                    "width": "100%",
                  },
                  undefined,
                  {},
                  {},
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        [
          {},
          {
            "gap": 16,
            "paddingBottom": 16,
            "position": "relative",
          },
          undefined,
          {},
          {},
        ]
      }
      withPadding={true}
    >
      <View
        style={
          [
            {
              "flex": 1,
            },
            {
              "elevation": 3,
              "height": 16,
              "left": 0,
              "marginTop": -16,
              "position": "absolute",
              "right": 0,
              "zIndex": 3,
            },
            undefined,
            {},
            {},
          ]
        }
      >
        <RNSVGSvgView
          align="none"
          bbHeight="100%"
          bbWidth="100%"
          focusable={false}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": "100%",
                "width": "100%",
              },
            ]
          }
          vbHeight={100}
          vbWidth={100}
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
          >
            <RNSVGDefs>
              <RNSVGLinearGradient
                gradient={
                  [
                    0,
                    16777215,
                    1,
                    -1,
                  ]
                }
                gradientTransform={
                  [
                    6.123233995736766e-17,
                    1,
                    -1,
                    6.123233995736766e-17,
                    0,
                    0,
                  ]
                }
                gradientUnits={1}
                name="g"
                x1="0"
                x2="100"
                y1="0"
                y2="0"
              />
            </RNSVGDefs>
            <RNSVGRect
              fill={
                {
                  "brushRef": "g",
                  "type": 1,
                }
              }
              height="100%"
              propList={
                [
                  "fill",
                ]
              }
              width="100%"
              x="0"
              y="0"
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
      <View
        accessibilityHint="Start"
        accessibilityLabel="Start"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        disableHaptic={false}
        focusable={true}
        hitSlop={10}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "height": 56,
              "position": "relative",
            },
            undefined,
            {},
          ]
        }
      >
        <View
          disabled={false}
          style={
            [
              {},
              {
                "backgroundColor": "#2e2e2e",
                "borderRadius": 8,
                "bottom": -5,
                "height": "100%",
                "left": 0,
                "position": "absolute",
                "width": "100%",
              },
              undefined,
              {},
              {},
            ]
          }
        />
        <View
          collapsable={false}
          color="$brand"
          disabled={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#2196f3",
              "borderColor": "#2e2e2e",
              "borderRadius": 8,
              "borderWidth": 2,
              "display": "flex",
              "flexDirection": "row",
              "height": "100%",
              "justifyContent": "space-between",
              "marginTop": 0,
              "paddingHorizontal": 16,
              "width": "100%",
            }
          }
        >
          <View
            disabled={false}
            hasAdornment={false}
            style={
              [
                {},
                {
                  "height": "100%",
                  "justifyContent": "center",
                  "opacity": 1,
                  "width": "100%",
                },
                undefined,
                {},
                {},
              ]
            }
          >
            <Text
              isLightBackground={false}
              style={
                [
                  {
                    "fontFamily": "BMJUA",
                    "fontSize": 16,
                    "paddingBottom": 4,
                  },
                  {
                    "color": "#ffffff",
                    "fontSize": 24,
                    "marginVertical": 0,
                    "paddingBottom": 4,
                    "textAlign": "center",
                  },
                  undefined,
                  {},
                  {},
                ]
              }
            >
              Start
            </Text>
          </View>
        </View>
        <View
          style={
            [
              {},
              {
                "borderRadius": 8,
                "bottom": -5,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              undefined,
              {},
              {},
            ]
          }
        >
          <View
            collapsable={false}
            isLightBackground={false}
            style={
              {
                "backgroundColor": "#ffffff",
                "borderRadius": 8,
                "height": "100%",
                "left": 0,
                "marginTop": 0,
                "opacity": 0.2,
                "position": "absolute",
                "width": "0%",
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
